Dependices libraries:


*   Numpy
*   pandas

*   sklearn --> model_selection --> train_test_split
*   sklearn --> linear_model --> LogisticRegression

*   sklearn --> metrics --> accuracy_score











In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [16]:
import os 
input_file = os.path.join(os.getcwd(),'sonar data.csv')
print(input_file)
os.listdir()

/home/shubhm/Desktop/Trivium/TCS project/sonar_rock_vs_mine_prediction/sonar data.csv


['Project_Rock_vs_Mine_prediction.ipynb',
 '.ipynb_checkpoints',
 'sonar data.csv',
 'sonar_weights.sav',
 'sonar_weights1.sav']

We can upload data in colab by clicking uplaod button.

In [17]:
# uploading data in pandas DataFrame
sonar_data = pd.read_csv(input_file, header = None)

In [18]:
sonar_data.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      51      52      53      54      55      56      57  \
0  0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1  0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2  0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3  0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4  0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   

       58      59  60  
0  0.0090  0.0032   R  
1  0.0052  0.0044   R  
2  0.0095  0.0078   R  
3  0.0040  0.0117   R  
4  0.0107  0.0094   R  

[5 rows x 61 columns]

The above dataset has 60 colums

> Last columns tell the prediction, thus we have 59 + 1 columns



In [19]:
# last 5 rows of the data
sonar_data.tail()

0       1       2       3       4       5       6       7       8   \
203  0.0187  0.0346  0.0168  0.0177  0.0393  0.1630  0.2028  0.1694  0.2328   
204  0.0323  0.0101  0.0298  0.0564  0.0760  0.0958  0.0990  0.1018  0.1030   
205  0.0522  0.0437  0.0180  0.0292  0.0351  0.1171  0.1257  0.1178  0.1258   
206  0.0303  0.0353  0.0490  0.0608  0.0167  0.1354  0.1465  0.1123  0.1945   
207  0.0260  0.0363  0.0136  0.0272  0.0214  0.0338  0.0655  0.1400  0.1843   

         9   ...      51      52      53      54      55      56      57  \
203  0.2684  ...  0.0116  0.0098  0.0199  0.0033  0.0101  0.0065  0.0115   
204  0.2154  ...  0.0061  0.0093  0.0135  0.0063  0.0063  0.0034  0.0032   
205  0.2529  ...  0.0160  0.0029  0.0051  0.0062  0.0089  0.0140  0.0138   
206  0.2354  ...  0.0086  0.0046  0.0126  0.0036  0.0035  0.0034  0.0079   
207  0.2354  ...  0.0146  0.0129  0.0047  0.0039  0.0061  0.0040  0.0036   

         58      59  60  
203  0.0193  0.0157   M  
204  0.0062  0.0067   M  
205  0.0077  0.0031   M  
206  0.0036  0.0048   M  
207  0.0061  0.0115   M  

[5 rows x 61 columns]

In [20]:
# shape of the data
sonar_data.shape

(208, 61)

In [21]:
# getting some insigits of stastical measures about the dataset
sonar_data.describe()

0           1           2           3           4           5   \
count  208.000000  208.000000  208.000000  208.000000  208.000000  208.000000   
mean     0.029164    0.038437    0.043832    0.053892    0.075202    0.104570   
std      0.022991    0.032960    0.038428    0.046528    0.055552    0.059105   
min      0.001500    0.000600    0.001500    0.005800    0.006700    0.010200   
25%      0.013350    0.016450    0.018950    0.024375    0.038050    0.067025   
50%      0.022800    0.030800    0.034300    0.044050    0.062500    0.092150   
75%      0.035550    0.047950    0.057950    0.064500    0.100275    0.134125   
max      0.137100    0.233900    0.305900    0.426400    0.401000    0.382300   

               6           7           8           9   ...          50  \
count  208.000000  208.000000  208.000000  208.000000  ...  208.000000   
mean     0.121747    0.134799    0.178003    0.208259  ...    0.016069   
std      0.061788    0.085152    0.118387    0.134416  ...    0.012008   
min      0.003300    0.005500    0.007500    0.011300  ...    0.000000   
25%      0.080900    0.080425    0.097025    0.111275  ...    0.008425   
50%      0.106950    0.112100    0.152250    0.182400  ...    0.013900   
75%      0.154000    0.169600    0.233425    0.268700  ...    0.020825   
max      0.372900    0.459000    0.682800    0.710600  ...    0.100400   

               51          52          53          54          55          56  \
count  208.000000  208.000000  208.000000  208.000000  208.000000  208.000000   
mean     0.013420    0.010709    0.010941    0.009290    0.008222    0.007820   
std      0.009634    0.007060    0.007301    0.007088    0.005736    0.005785   
min      0.000800    0.000500    0.001000    0.000600    0.000400    0.000300   
25%      0.007275    0.005075    0.005375    0.004150    0.004400    0.003700   
50%      0.011400    0.009550    0.009300    0.007500    0.006850    0.005950   
75%      0.016725    0.014900    0.014500    0.012100    0.010575    0.010425   
max      0.070900    0.039000    0.035200    0.044700    0.039400    0.035500   

               57          58          59  
count  208.000000  208.000000  208.000000  
mean     0.007949    0.007941    0.006507  
std      0.006470    0.006181    0.005031  
min      0.000300    0.000100    0.000600  
25%      0.003600    0.003675    0.003100  
50%      0.005800    0.006400    0.005300  
75%      0.010350    0.010325    0.008525  
max      0.044000    0.036400    0.043900  

[8 rows x 60 columns]

In [22]:
# type of data of 60th column of the dataset
sonar_data[60].value_counts()

M    111
R     97
Name: 60, dtype: int64

Thus the 60th colums has 111 --> Mine and 97 --> Rock. The dataset almost eqaul.

In [23]:
# let sort the data based on the prediction classes Rock vs Mine
sonar_data.groupby(60).mean()

0         1         2         3         4         5         6   \
60                                                                         
M   0.034989  0.045544  0.050720  0.064768  0.086715  0.111864  0.128359   
R   0.022498  0.030303  0.035951  0.041447  0.062028  0.096224  0.114180   

          7         8         9   ...        50        51        52        53  \
60                                ...                                           
M   0.149832  0.213492  0.251022  ...  0.019352  0.016014  0.011643  0.012185   
R   0.117596  0.137392  0.159325  ...  0.012311  0.010453  0.009640  0.009518   

          54        55        56        57        58        59  
60                                                              
M   0.009923  0.008914  0.007825  0.009060  0.008695  0.006930  
R   0.008567  0.007430  0.007814  0.006677  0.007078  0.006024  

[2 rows x 60 columns]

If we campare the mean values of the above classes, we can quit difference between the dataset. This import to see during making model.

In [24]:
# sperating and lebeling the data, axis =0 mean row and axis = 1 means columns
X = sonar_data.drop(columns = 60, axis = 1)
Y= sonar_data[60]

In [25]:
print(X)
print(Y)

         0       1       2       3       4       5       6       7       8   \
0    0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1    0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2    0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3    0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4    0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
203  0.0187  0.0346  0.0168  0.0177  0.0393  0.1630  0.2028  0.1694  0.2328   
204  0.0323  0.0101  0.0298  0.0564  0.0760  0.0958  0.0990  0.1018  0.1030   
205  0.0522  0.0437  0.0180  0.0292  0.0351  0.1171  0.1257  0.1178  0.1258   
206  0.0303  0.0353  0.0490  0.0608  0.0167  0.1354  0.1465  0.1123  0.1945   
207  0.0260  0.0363  0.0136  0.0272  0.0214  0.0338  0.0655  0.1400  0.1843   

         9   ...      50      51      52      53   

Now spliting data in train and test data 

In [69]:
# Here stratify Y means, data equally slpit based on Y class and random_state means data randomly split on random state 1 (it could be split random state 1,2,3...)
X_train, X_test, Y_train, Y_test = train_test_split(X,Y,test_size = 0.1, stratify = Y, random_state = 2)

In [70]:
# checking size of data
print(X.shape, X_train.shape,X_test.shape)

(208, 60) (187, 60) (21, 60)


In [71]:
print(Y.shape, Y_train.shape,Y_test.shape)

(208,) (187,) (21,)


In [72]:
print(X_train)
print(Y_train)

         0       1       2       3       4       5       6       7       8   \
113  0.0283  0.0599  0.0656  0.0229  0.0839  0.1673  0.1154  0.1098  0.1370   
25   0.0201  0.0026  0.0138  0.0062  0.0133  0.0151  0.0541  0.0210  0.0505   
116  0.0094  0.0333  0.0306  0.0376  0.1296  0.1795  0.1909  0.1692  0.1870   
171  0.0179  0.0136  0.0408  0.0633  0.0596  0.0808  0.2090  0.3465  0.5276   
64   0.0071  0.0103  0.0135  0.0494  0.0253  0.0806  0.0701  0.0738  0.0117   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
85   0.0365  0.1632  0.1636  0.1421  0.1130  0.1306  0.2112  0.2268  0.2992   
132  0.0968  0.0821  0.0629  0.0608  0.0617  0.1207  0.0944  0.4223  0.5744   
164  0.0163  0.0198  0.0202  0.0386  0.0752  0.1444  0.1487  0.1484  0.2442   
19   0.0126  0.0149  0.0641  0.1732  0.2565  0.2559  0.2947  0.4110  0.4983   
37   0.0333  0.0221  0.0270  0.0481  0.0679  0.0981  0.0843  0.1172  0.0759   

         9   ...      50      51      52      53   

After getting clean data, Now let train our dataset by Logisitc Regression


In [73]:
# saving our Logisitic Regression method into the varible named model
model = LogisticRegression()

In [74]:
# fiting the data 
model.fit(X_train,Y_train)

LogisticRegression()

Model accuracy

In [75]:
# accuracy on the training data
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [76]:
print('Accuracy on training data: ',training_data_accuracy)

Accuracy on training data:  0.8128342245989305


In [78]:
# accuracy on the test data
X_test_prediction  = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction,Y_test)

In [79]:
print('Accuracy on the test data: ', test_data_accuracy)

Accuracy on the test data:  0.9047619047619048


Our model has been prepared and it has train accuracy around 80% and test accuracy 76 %, which is good prediction.

Making predictive system of the prediction


In [80]:
input_data = (0.0158,0.0239,0.0150,0.0494,0.0988,0.1425,0.1463,0.1219,0.1697,0.1923,0.2361,0.2719,0.3049,0.2986,0.2226,0.1745,0.2459,0.3100,0.3572,0.4283,0.4268,0.3735,0.4585,0.6094,0.7221,0.7595,0.8706,1.0000,0.9815,0.7187,0.5848,0.4192,0.3756,0.3263,0.1944,0.1394,0.1670,0.1275,0.1666,0.2574,0.2258,0.2777,0.1613,0.1335,0.1976,0.1234,0.1554,0.1057,0.0490,0.0097,0.0223,0.0121,0.0108,0.0057,0.0028,0.0079,0.0034,0.0046,0.0022,0.0021)
#chaning data into numpy array
input_data_as_numpy_array = np.asarray(input_data)

# reshaping data as we are predicting as on instance
input_data_reshaped = input_data_as_numpy_array.reshape(1,-1)

# preditction
prediction = model.predict(input_data_reshaped)
print(prediction)

#############
if prediction[0] == 'R':
  print('The object is a Rock.')
else:
  print('The object is a Mine.')

['M']
The object is a Mine.


# Saving weights of the trained model data.
To save our machine learning model can be save many ways:


1.   Save Your Model with pickle
2.   Save Your Model with joblib



In [81]:
# Saving model with pickle
import pickle
data_weights = 'sonar_weights_new.sav'
pickle.dump(model,open(data_weights, 'wb'))


In [82]:
# reusing this weights file to make prediction
loaded_model = pickle.load(open('sonar_weights_new.sav', 'rb'))
result = loaded_model.predict(input_data_reshaped)
print(result)

['M']


In [56]:
# saving our model with joblib
import joblib
joblib.dump(model,'sonar_weights1.sav')

['sonar_weights1.sav']

In [57]:
# reusing this weights file to make prediction
loaded_model1 = joblib.load('sonar_weights1.sav')
result1 = loaded_model1.predict(input_data_reshaped)
print(result1)

['M']
